### DISCUSSION TITLE
defining an object in global variables and marking keys as readonly

### DISCUSSION THREAD

**Discussion by gaurav5430** [eslint/eslint#15220] [Q&A]
i have been using global variables in eslintrc to make sure my global variables that are injected through webpack are identified by eslint and do not lead to a no-undef error. Another way this helps is that it prevents writing to those globals by mistake (as they are defined as readonly). This works fine with defining top level keys directly as readonly, but I can't seem to find a way to define an object with readonly keys.

what I would like to have is:

```
OBJ: { 
  KEY1: "readonly",
  KEY2: "readonly",
}
```

so that if someone tries to use `OBJ.KEY3` eslint gives an error for undefined global

or if someone tries to assign `OBJ.KEY1 = 'somevalue'` eslint gives an error.

is this possible?
does it even make sense to use eslint globals for this?

( i can flatten my object and define each key separately as a workaround, but having an object helps me group related globals)

**Comment by nzakas** [comment]
You would need to create a custom rule for that. ESLint globals aren’t intended for this use case.

**Reply by gaurav5430** [reply]
ok, so are globals only intended for eslint to identify the top level key and mark it read-only if required? 

just trying to understand, what would be the reasoning behind limiting it to just one level?

**Reply by tonyadudas** [reply]
Hey

**Reply by nzakas** [reply]
Because global variables are special in JavaScript. It’s easy to accidentally overwrite them. Chances are you won’t accidentally overwrite a property assignment.

but again, you can always write a custom rule if you want to do that.
